contracts/DKVDaggerHashimoto.sol

```diff
@@ -212,7 +216,66 @@ contract DecentralizedKVDaggerHashimoto is DecentralizedKV {
             uint256 parent = uint256(hash0) % rows;
             uint256 kvIdx = parent + (startShardId << shardEntryBits);
             bytes memory data = maskedData[i];
-            require(systemContract.checkDaggerData(kvIdx, kvMap[idxMap[kvIdx]].hash, data), "invalid access proof");
+            require(systemContract.checkDaggerData(kvIdx, kvMap[idxMap[kvIdx]].hash, data),
+                    "invalid access proof");
+
+            assembly {
+                mstore(data, hash0)
+                hash0 := keccak256(data, add(maxKvSize, 0x20))
+                mstore(data, maxKvSize)
+            }
+        }
+        return hash0;
+    }
+
+    /* Helper function to avoid stack to deep error */
+    function _generateKVIdx(
+        bytes32 hash0,
+        uint256 startShardId,
+        uint256 shardLenBits
+    ) internal view returns (uint256) {
+        uint256 rows = 1 << (shardEntryBits + shardLenBits);
+        uint256 parent = uint256(hash0) % rows;
+        uint256 kvIdx = parent + (startShardId << shardEntryBits);
+        return kvIdx;
+    }
+
+    function _generateChunkIdx(
+        bytes32 hash0
+    ) internal view returns (uint256) {
+        require((CHUNK_SIZE != 0) && ((CHUNK_SIZE & (CHUNK_SIZE - 1)) == 0), "CHUNK_SIZE invalid");
+        uint256 maxKvSize = 1 << maxKvSizeBits;
+        uint256 chunksPerKVBlock = maxKvSize / CHUNK_SIZE;
+        /* If maxKvSize is smaller than CHUNK_SIZE, early exit */
+        if (chunksPerKVBlock == 0) return 0;
```

Comment:
This should never happen - or we could put it as immutable so that we don't need to evaluate the number per call.

Reply:
In fact, it can happen, at least in our test env. 
In test env we have maxKvSize around 64B - 128B, smaller than default CHUNK_SIZE 4k.
Under above situation, Hashimoto with Merkle proof degenerates one simple keccak256. 
But the corner case remains.

Reply:
Should not in that case that CHUNK_SIZE = 64B or 128B?

Reply:
Currently we make CHUNK_SIZE a constant but yep I guess that's basically a style issue. Perhaps we can change that.

Reply:
Yes, we should also stick to the assumption that max value size is a 2^n multiple of CHUNK_SIZE.  Further, we should minimize the checks in the contract call methods that can be called frequently.